TestConvertResources/backend-lb-policy flake

https://prow.istio.io/view/gs/istio-prow/logs/unit-tests-arm64_istio_postsubmit/2001348563858100224

This is still flaking

I've tried it with `stress` and reverting #58590 and it still shows up. 

```
--- FAIL: TestConvertResources (1.73s)
    --- FAIL: TestConvertResources/mix-backend-policy (0.37s)
        conversion_test.go:809: Failed validating golden file testdata/mix-backend-policy.yaml.golden:
            @@ -1,16 +1 @@
            -apiVersion: networking.istio.io/v1
            -kind: DestinationRule
            -metadata:
            -  annotations:
            -    internal.istio.io/parents: XBackendTrafficPolicy/default.lb-policy,BackendTLSPolicy/default.tls-upstream-echo
            -  name: echo~istio-autogenerated-k8s-gateway
            -  namespace: default
            -spec:
            -  host: echo.default.svc.domain.suffix
            -  trafficPolicy:
            -    loadBalancer: {}
            -    tls:
            -      credentialName: configmap://default/auth-cert
            -      mode: SIMPLE
            -      sni: auth.example.com
            ----
            +
        conversion_test.go:843: - name: gateway/tagWatcher
              state:
                outputs:
                  static: 1
                synced: true
              uid: 4
            - name: gateway/gatewayContext
              state:
                outputs:
                  static: 1
                synced: true
              uid: 5
            - name: gateway/informer/Namespaces
              state:
                outputs:
                  default:
                    apiVersion: v1
                    kind: Namespace
                    metadata:
```

--- a/pilot/pkg/config/kube/gateway/backend_policies.go
+++ b/pilot/pkg/config/kube/gateway/backend_policies.go
@@ -237,7 +237,7 @@ func DestinationRuleCollection(
 					}
 				}
 			case kind.ServiceEntry:
-				serviceEntryObj, err := references.LocalPolicyTargetRef(gw.LocalPolicyTargetReference{
+				serviceEntryObj, err := references.LocalPolicyTargetRef(ctx, gw.LocalPolicyTargetReference{
 					Group: "networking.istio.io",
 					Kind:  "ServiceEntry",
 					Name:  gw.ObjectName(target.Name),
@@ -316,7 +316,7 @@ func BackendTLSPolicyCollection(
 			}
 			return nil
 		})
-		tls.CredentialName = getBackendTLSCredentialName(s.Validation, i.Namespace, conds, references)
+		tls.CredentialName = getBackendTLSCredentialName(ctx, s.Validation, i.Namespace, conds, references)
 
 		// In ancestor status, we need to report for Service (for mesh) and for each relevant Gateway.
 		// However, there is a max of 16 items we can report.
@@ -328,7 +328,7 @@ func BackendTLSPolicyCollection(
 		uniqueGateways := sets.New[types.NamespacedName]()
 		for idx, t := range i.Spec.TargetRefs {
 			conds = maps.Clone(conds)
-			refo, err := references.LocalPolicyTargetRef(t.LocalPolicyTargetReference, i.Namespace)
+			refo, err := references.LocalPolicyTargetRef(ctx, t.LocalPolicyTargetReference, i.Namespace)
 			var sectionName *string
 			if err == nil {
 				switch refType := refo.(type) {
@@ -432,6 +432,7 @@ func BackendTLSPolicyCollection(
 }
 
 func getBackendTLSCredentialName(
+	ctx krt.HandlerContext,
 	validation gw.BackendTLSPolicyValidation,
 	policyNamespace string,
 	conds map[string]*condition,
@@ -459,7 +460,7 @@ func getBackendTLSCredentialName(
 	if len(validation.CACertificateRefs) > 1 {
 		conds[string(gw.PolicyConditionAccepted)].message += "; warning: only the first caCertificateRefs will be used"
 	}
-	refo, err := references.LocalPolicyRef(ref, policyNamespace)
+	refo, err := references.LocalPolicyRef(ctx, ref, policyNamespace)
 	if err == nil {
 		switch to := refo.(type) {
 		case *v1.ConfigMap:
@@ -556,7 +557,7 @@ func BackendTrafficPolicyCollection(
 
 		for idx, t := range i.Spec.TargetRefs {
 			conds = maps.Clone(conds)
-			refo, err := references.XLocalPolicyTargetRef(t, i.Namespace)
+			refo, err := references.XLocalPolicyTargetRef(ctx, t, i.Namespace)
 			if err == nil {
 				switch refo.(type) {
 				case *v1.Service:

--- a/pilot/pkg/config/kube/gateway/references.go
+++ b/pilot/pkg/config/kube/gateway/references.go
@@ -29,35 +29,35 @@ import (
 // ReferenceSet stores a variety of different types of resource, and allows looking them up as Gateway API references.
 // This is merely a convenience to avoid needing to lookup up a bunch of types all over the place.
 type ReferenceSet struct {
-	erasedCollections map[config.GroupVersionKind]func(name, namespace string) (any, bool)
+	erasedCollections map[config.GroupVersionKind]func(ctx krt.HandlerContext, name, namespace string) (any, bool)
 }
 
-func (s ReferenceSet) LocalPolicyTargetRef(ref gatewayv1.LocalPolicyTargetReference, localNamespace string) (any, error) {
-	return s.internal(string(ref.Name), string(ref.Group), string(ref.Kind), localNamespace)
+func (s ReferenceSet) LocalPolicyTargetRef(ctx krt.HandlerContext, ref gatewayv1.LocalPolicyTargetReference, localNamespace string) (any, error) {
+	return s.internal(ctx, string(ref.Name), string(ref.Group), string(ref.Kind), localNamespace)
 }
 
-func (s ReferenceSet) XLocalPolicyTargetRef(ref gatewayx.LocalPolicyTargetReference, localNamespace string) (any, error) {
-	return s.internal(string(ref.Name), string(ref.Group), string(ref.Kind), localNamespace)
+func (s ReferenceSet) XLocalPolicyTargetRef(ctx krt.HandlerContext, ref gatewayx.LocalPolicyTargetReference, localNamespace string) (any, error) {
+	return s.internal(ctx, string(ref.Name), string(ref.Group), string(ref.Kind), localNamespace)
 }
 
-func (s ReferenceSet) LocalPolicyRef(ref gatewayv1.LocalObjectReference, localNamespace string) (any, error) {
-	return s.internal(string(ref.Name), string(ref.Group), string(ref.Kind), localNamespace)
+func (s ReferenceSet) LocalPolicyRef(ctx krt.HandlerContext, ref gatewayv1.LocalObjectReference, localNamespace string) (any, error) {
+	return s.internal(ctx, string(ref.Name), string(ref.Group), string(ref.Kind), localNamespace)
 }
 
-func (s ReferenceSet) internal(name, group, kind, localNamespace string) (any, error) {
+func (s ReferenceSet) internal(ctx krt.HandlerContext, name, group, kind, localNamespace string) (any, error) {
 	t := normalizeReference(&group, &kind, config.GroupVersionKind{})
 	lookup, f := s.erasedCollections[t]
 	if !f {
 		return nil, fmt.Errorf("unsupported kind %v", kind)
 	}
-	if v, ok := lookup(name, localNamespace); ok {
+	if v, ok := lookup(ctx, name, localNamespace); ok {
 		return v, nil
 	}
 	return nil, fmt.Errorf("reference %v/%v (of kind %v) not found", localNamespace, name, kind)
 }
 
 func NewReferenceSet(opts ...func(r *ReferenceSet)) *ReferenceSet {
-	r := &ReferenceSet{erasedCollections: make(map[config.GroupVersionKind]func(name, namespace string) (any, bool))}
+	r := &ReferenceSet{erasedCollections: make(map[config.GroupVersionKind]func(ctx krt.HandlerContext, name, namespace string) (any, bool))}
 	for _, opt := range opts {
 		opt(r)
 	}
@@ -67,8 +67,8 @@ func NewReferenceSet(opts ...func(r *ReferenceSet)) *ReferenceSet {
 func AddReference[T runtime.Object](c krt.Collection[T]) func(r *ReferenceSet) {
 	return func(r *ReferenceSet) {
 		g := schematypes.MustGVKFromType[T]()
-		r.erasedCollections[g] = func(name, namespace string) (any, bool) {
-			o := c.GetKey(namespace + "/" + name)
+		r.erasedCollections[g] = func(ctx krt.HandlerContext, name, namespace string) (any, bool) {
+			o := krt.FetchOne(ctx, c, krt.FilterKey(namespace+"/"+name))
 			if o == nil {
 				return nil, false
 			}

